src/v/cloud_roles/aws_sts_refresh_impl.cc

```diff
@@ -0,0 +1,166 @@
+    static constexpr std::string_view access_key_id = "AccessKeyId";
+    static constexpr std::string_view secret_access_key = "SecretAccessKey";
+    static constexpr std::string_view session_token = "SessionToken";
+};
+
+static const boost::beast::string_view content_type{
+  "application/x-www-form-urlencoded"};
+
+static constexpr std::string_view request_payload
+  = "Action=AssumeRoleWithWebIdentity"
+    "&DurationSeconds={}"
+    "&Version={}"
+    "&RoleSessionName={}"
+    "&RoleArn={}"
+    "&WebIdentityToken={}";
+
+static ss::sstring load_from_env(std::string_view env_var) {
+    auto env_value = std::getenv(env_var.data());
+    if (!env_value) {
+        throw std::runtime_error(fmt::format(
+          "environment variable {} is not set, the STS client cannot function "
+          "without this.",
+          env_var));
+    }
+    return env_value;
+}
+
+cloud_roles::aws_sts_refresh_impl::aws_sts_refresh_impl(
+  seastar::sstring api_host,
+  uint16_t api_port,
+  s3::aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    std::move(api_host), api_port, std::move(region), as, retry_params)
+  , _role{load_from_env(aws_injected_env_vars::role_arn)}
+  , _token_file_path{load_from_env(aws_injected_env_vars::token_file_path)} {}
+
+ss::future<cloud_roles::api_response>
+cloud_roles::aws_sts_refresh_impl::fetch_credentials() {
+    // The content of the token file is periodically rotated by k8s, so we
+    // read it fully each time to avoid stale tokens
+
+    ss::sstring token{};
+    try {
+        token = co_await read_fully_to_string({_token_file_path});
+    } catch (...) {
+        vlog(
+          clrl_log.error,
+          "failed to read IRSA pod token from file {}",
+          _token_file_path);
+        throw;
+    }
+
+    boost::beast::http::request<boost::beast::http::string_body> assume_req;
+    assume_req.method(boost::beast::http::verb::post);
+    assume_req.target("/");
+
+    assume_req.set(boost::beast::http::field::content_type, content_type);
+
+    ss::sstring body = fmt::format(
+      request_payload,
+      sts_params::token_expiry_seconds.count(),
+      sts_params::api_version,
+      sts_params::role_session_name,
+      _role,
+      token);
+
+    co_return co_await post_request(
+      make_api_client(), std::move(assume_req), std::move(body));
+}
+
+cloud_roles::api_response_parse_result
+cloud_roles::aws_sts_refresh_impl::parse_response(iobuf resp) {
+    auto root = s3::iobuf_to_ptree(std::move(resp));
+
+    auto creds_node_maybe = root.get_child_optional(
+      response_node_credential_path.data());
+
+    if (!creds_node_maybe) {
+        return malformed_api_response_error{
+          .missing_fields = {response_node_credential_path.data()}};
+    }
+
+    auto creds_node = *creds_node_maybe;
+
+    std::vector<ss::sstring> missing_fields;
+    for (const auto& key :
+         {sts_response_schema::expiry,
```

Comment:
+1 super nice, going to use this pattern myself

Reply:
I didn't mention it here, but really for these cases--assuming the keys aren't dynamic and it doesn't look that way--rapidjson has schema validation. examples of how to use are in admin_server.cc.